graphql_compiler/compiler/ir_lowering_sql/__init__.py

```diff
@@ -0,0 +1,264 @@
+
+    Args:
+        ir_blocks: list of IR blocks to lower into SQL-compatible form
+        query_metadata_table: QueryMetadataTable object containing all metadata collected during
+                              query processing, including location metadata (e.g. which locations
+                              are folded or optional).
+        type_equivalence_hints: optional dict of GraphQL interface or type -> GraphQL union.
+                                Used as a workaround for GraphQL's lack of support for
+                                inheritance across "types" (i.e. non-interfaces), as well as a
+                                workaround for Gremlin's total lack of inheritance-awareness.
+                                The key-value pairs in the dict specify that the "key" type
+                                is equivalent to the "value" type, i.e. that the GraphQL type or
+                                interface in the key is the most-derived common supertype
+                                of every GraphQL type in the "value" GraphQL union.
+                                Recursive expansion of type equivalence hints is not performed,
+                                and only type-level correctness of this argument is enforced.
+                                See README.md for more details on everything this parameter does.
+                                *****
+                                Be very careful with this option, as bad input here will
+                                lead to incorrect output queries being generated.
+                                *****
+
+    Returns:
+        SqlTree object containing SqlNodes organized in a tree structure.
+    """
+    query_path_to_location_info = {
+        location.query_path: location_info
+        for location, location_info in query_metadata_table.registered_locations
+    }
+    # Perform lowering passes over IR blocks
+    construct_result = ir_blocks.pop()
+    construct_result = lower_construct_result(construct_result)
+    block_index_to_location = get_block_index_to_location_map(ir_blocks)
+    ir_blocks = lower_unary_transformations(ir_blocks)
+    ir_blocks = lower_context_field_existence(ir_blocks)
+    ir_blocks = lower_optional_fields(
+        ir_blocks, block_index_to_location, query_path_to_location_info)
+    query_path_to_node = {}
+    query_path_to_filter = defaultdict(list)
+    query_path_to_tag_fields = defaultdict(list)
+    tree_root = None
+    for index, block in enumerate(ir_blocks):
+        if isinstance(block, (blocks.Recurse, blocks.Traverse, blocks.QueryRoot)):
+            location = block_index_to_location[index]
+            location_info = query_metadata_table.get_location_info(location)
+            query_path = location.query_path
+            if tree_root is None:
+                if not isinstance(block, blocks.QueryRoot):
+                    raise AssertionError(u'Encountered QueryRoot but tree root is already set.')
```

Comment:
In error messages, always include the values based on which you are raising the error -- in this case, the `block` and the `tree_root`. That will make getting to the bottom of the error much easier.

In this case, I might also include the `ir_blocks` and the `query_metadata_table` because having more data is always helpful.